src/Database/PostgreSQL/Simple/Internal.hs

```diff
@@ -329,40 +330,70 @@ exec conn sql =
     withConnection conn $ \h -> do
         success <- PQ.sendQuery h sql
         if success
-        then awaitResult h Nothing
+        then do
+          mfd <- PQ.socket h
+          case mfd of
+            Nothing -> throwIO $! fdError "Database.PostgreSQL.Simple.Internal.exec"
+            Just socket ->
+              -- Here we assume any exceptions are asynchronous, or that
+              -- they are not from libpq. If an error happens in libpq
+              -- (e.g. the query being canceled or session terminated),
+              -- libpq will not throw, but will instead return a Result
+              -- indicating an error
+              uninterruptibleMask $ \restore ->
```

Comment:
I think you might get an async exception after `sendQuery` but before `uninterruptibleMask`. Probably the whole `do` block needs `mask` around it with  restore around `sendQuery`. Sorry if I'm overlooking something here.

Reply:
Ah, good point, thanks!